redis/lookup.go

```diff
@@ -0,0 +1,216 @@
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/letsencrypt/boulder/cmd"
+	blog "github.com/letsencrypt/boulder/log"
+
+	"github.com/redis/go-redis/v9"
+)
+
+var ErrNoShardsResolved = errors.New("0 shards were resolved")
+
+// Lookup wraps a Redis ring client by reference and keeps the Redis ring shards
+// up to date via periodic SRV lookups.
+type Lookup struct {
+	// srvLookups is a list of SRV records to be looked up.
+	srvLookups []cmd.ServiceDomain
+
+	// updateFrequency is the frequency of periodic SRV lookups. Defaults to 30
+	// seconds.
+	updateFrequency time.Duration
+
+	// updateTimeout is the timeout for each SRV lookup. Defaults to 90% of the
+	// update frequency.
+	updateTimeout time.Duration
+
+	// dnsAuthority is the single <hostname|IPv4|[IPv6]>:<port> of the DNS
+	// server to be used for SRV lookups. If the address contains a hostname it
+	// will be resolved via the system DNS. If the port is left unspecified it
+	// will default to '53'. If this field is left unspecified the system DNS
+	// will be used for resolution.
+	dnsAuthority string
+
+	resolver *net.Resolver
+	ring     *redis.Ring
+	logger   blog.Logger
+}
+
+// NewLookup constructs and returns a new Lookup instance. An initial SRV lookup
+// is performed to populate the Redis ring shards. If this lookup fails or
+// otherwise results in an empty set of resolved shards, an error is returned.
+func NewLookup(srvLookups []cmd.ServiceDomain, dnsAuthority string, frequency time.Duration, ring *redis.Ring, logger blog.Logger) (*Lookup, error) {
+	updateFrequency := frequency
+	if updateFrequency <= 0 {
+		// Set default frequency.
+		updateFrequency = 30 * time.Second
+	}
+	// Set default timeout to 90% of the update frequency.
+	updateTimeout := updateFrequency - updateFrequency/10
+
+	var lookup *Lookup
+	if dnsAuthority == "" {
+		// Use the system DNS resolver.
+		lookup = &Lookup{
+			srvLookups:      srvLookups,
+			ring:            ring,
+			logger:          logger,
+			updateFrequency: updateFrequency,
+			updateTimeout:   updateTimeout,
+			resolver:        net.DefaultResolver,
+			dnsAuthority:    dnsAuthority,
+		}
+	} else {
+		// Setup a custom DNS resolver.
+		lookup = &Lookup{
+			srvLookups:      srvLookups,
+			ring:            ring,
+			logger:          logger,
+			updateFrequency: updateFrequency,
+			updateTimeout:   updateTimeout,
+			dnsAuthority:    dnsAuthority,
+		}
+
+		host, port, err := net.SplitHostPort(dnsAuthority)
+		if err != nil {
+			// Assume only hostname or IPv4 address was specified.
+			host = dnsAuthority
+			port = "53"
+		}
+		lookup.dnsAuthority = net.JoinHostPort(host, port)
+		lookup.resolver = &net.Resolver{
+			PreferGo: true,
+			Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
+				// The custom resolver closes over the lookup.dnsAuthority field
+				// so it can be swapped out in testing.
+				return net.Dial(network, lookup.dnsAuthority)
+			},
+		}
+	}
```

Comment:
```suggestion
	var resolver *net.Resolver
	if dnsAuthority == "" {
		// Use the system DNS resolver.
		resolver = net.DefaultResolver
	} else {
		host, port, err := net.SplitHostPort(dnsAuthority)
		if err != nil {
			// Assume only hostname or IPv4 address was specified.
			host = dnsAuthority
			port = "53"
		}
		
		dnsAuthority = net.JoinHostPort(host, port)
		resolver = &net.Resolver{
			PreferGo: true,
			Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
				// The custom resolver closes over the lookup.dnsAuthority field
				// so it can be swapped out in testing.
				return net.Dial(network, lookup.dnsAuthority)
			},
		}
	}

	lookup = &Lookup{
		srvLookups:      srvLookups,
		ring:            ring,
		logger:          logger,
		updateFrequency: updateFrequency,
		updateTimeout:   updateTimeout,
		resolver:        resolver,
		dnsAuthority:    dnsAuthority,
	}
```

Reply:
Addressed in #7084.